### PR TITLE
Expose MiqEvent/EmsEvent group/levels in event stream OPTIONS

### LINKED ISSUE #1090: Api for timeline charts.
We have to convert existing pf charts to carbon. carbon doesn't have form elements like pf chart. But, we can create those using carbon react components and based on form results when apply clicked, we can display chart.

Screen Shot 2021-10-25 at 2 44 14 PM

Screen Shot 2021-10-25 at 4 17 29 PM

EDIT:

Adding task items:

- [x] core: https://github.com/ManageIQ/manageiq/pull/21552
- [x] core fix: https://github.com/ManageIQ/manageiq/pull/21569
  - [ ] followup - Fix (or add?) i18n to the backend _or_ on the API side: @jrafanie.
- [x] api: https://github.com/ManageIQ/manageiq-api/pull/1099
- [x] Test that you can filter event_streams based on group_names/group_levels which are different for each EventStream subclass.  For example, group names such as ems_operation/host_operations/compliance (MiqEvent) in addition to the EmsEvent ones: addition/configuration/console.  Same for group levels: sucess/failure (MiqEvent) and critical/detail/warning (EmsEvent). EDIT: Works: see https://github.com/ManageIQ/manageiq-api/issues/1090#issuecomment-974490596
- [ ] Add subcollection support for events streams in the primary collections that have it as an association.

### PR DESCRIPTION
**Pull request by jrafanie**
Part of #1090

Depends on: 
- [x] https://github.com/ManageIQ/manageiq/pull/21552
- [x] https://github.com/ManageIQ/manageiq/pull/21569
- [x] Write some tests once core PR is done.

OPTIONS http://localhost:3000/api/event_streams

RESPONSE:

```
{
    "attributes": [
        "availability_zone_id",
        "chain_id",
        "container_group_id",
        "container_group_name",
        "container_id",
        "container_name",
        "container_namespace",
        "container_node_id",
        "container_node_name",
        "container_replicator_id",
        "container_replicator_name",
        "created_on",
        "dest_ems_cluster_id",
        "dest_ems_cluster_name",
        "dest_ems_cluster_uid",
        "dest_host_id",
        "dest_host_name",
        "dest_vm_ems_ref",
        "dest_vm_location",
        "dest_vm_name",
        "dest_vm_or_template_id",
        "ems_cluster_id",
        "ems_cluster_name",
        "ems_cluster_uid",
        "ems_id",
        "ems_ref",
        "event_type",
        "full_data",
        "generating_ems_id",
        "group_id",
        "host_id",
        "host_name",
        "id",
        "is_task",
        "message",
        "physical_chassis_id",
        "physical_server_id",
        "physical_switch_id",
        "source",
        "target_id",
        "target_type",
        "tenant_id",
        "timestamp",
        "type",
        "user_id",
        "username",
        "vm_ems_ref",
        "vm_location",
        "vm_name",
        "vm_or_template_id"
    ],
    "virtual_attributes": [
        "group",
        "group_level",
        "group_name",
        "href_slug",
        "region_description",
        "region_number"
    ],
    "relationships": [
        "availability_zone",
        "container_group",
        "container_node",
        "container_replicator",
        "dest_host",
        "dest_miq_template",
        "dest_vm",
        "dest_vm_or_template",
        "ext_management_system",
        "generating_ems",
        "host",
        "miq_template",
        "physical_chassis",
        "physical_server",
        "physical_switch",
        "service",
        "target",
        "vm",
        "vm_or_template"
    ],
    "subcollections": [],
    "data": {
        "timeline_events": {
            "EmsEvent": {
                "description": "Management Events",
                "group_names": {
                    "other": "Other",
                    "addition": "Creation/Addition",
                    "configuration": "Configuration/Reconfiguration",
                    "console": "Console Activity",
                    "deletion": "Deletion/Removal",
                    "general": "General Activity",
                    "import_export": "Import/Export",
                    "migration": "Migration/Vmotion",
                    "network": "Network",
                    "power": "Power Activity",
                    "snapshot": "Snapshot Activity",
                    "status": "Status",
                    "storage": "Storage",
                    "update": "Update",
                    "firmware": "Firmware",
                    "devices": "Devices",
                    "login": "Login",
                    "security": "Security"
                },
                "group_levels": {
                    "critical": "Critical",
                    "detail": "Detail",
                    "warning": "Warning"
                }
            },
            "MiqEvent": {
                "description": "Policy Events",
                "group_names": {
                    "other": "Other",
                    "ems_operations": "Provider Operation",
                    "host_operations": "Host Operation",
                    "compliance": "Compliance",
                    "tags": "Company Tag",
                    "vm_operational": "VM Operation",
                    "vm_configurational": "VM Configuration",
                    "vm_process": "VM Lifecycle",
                    "service_process": "Service Lifecycle",
                    "orchestration_process": "Orchestration Lifecycle",
                    "storage_operational": "Datastore Operation",
                    "container_operations": "Container Operation",
                    "physical_server_operations": "Physical Server Operation",
                    "evm_operations": "Appliance Operation",
                    "middleware_operations": "Middleware Operation",
                    "auth_validation": "Authentication Validation (Provider)",
                    "authentication": "Authentication Validation"
                },
                "group_levels": {
                    "success": "Success",
                    "failure": "Failure",
                    "detail": "Detail"
                }
            }
        }
    }
}
```

### CODE REVIEW

**Comment by Fryguy** [comment]
@jrafanie I'm wonder if we should nest these options one more level under data.  That is, if we need to something else to OPTIONS in the future, the root of the data payload is taken.  Maybe:

```diff
      "data": {
-         "EmsEvent": {
+         "event_groups": {
+             "EmsEvent": {
```

(not sure on the name `event_groups`...we can debate that)

**Comment by jrafanie** [comment]
> Not sure if the "data" nesting belongs here or in the backend PR. I feel like the backend method should return the hash startng at the "EmsEvent" level, but the API should be the one that constructs the `data/event_groups` levels

I agree.  I'll nest this in this PR.  I like `timeline_events` as it summarizes the data below it... 

```
    "data": {
        "timeline_events": {
            "EmsEvent": {
                "description": "Management Events",
                "group_names": {
                    "addition": "Creation/Addition",
                    ...
                    "security": "Security"
                },
                "group_levels": {
                    "critical": "Critical",
                    "detail": "Detail",
                    "warning": "Warning"
                }
            },
            "MiqEvent": {
                "description": "Policy Events",
                "group_names": {
                    "ems_operations": "Provider Operation",
                    ...
                    "authentication": "Authentication Validation"
                },
                "group_levels": {
                    "success": "Success",
                    "failure": "Failure"
                }
            }
        }
    }
```

**Comment by jrafanie** [comment]
I have appeased rubocop but 😡 

image

File: app/controllers/api/event_streams_controller.rb [hunk 1]
```diff
@@ -1,4 +1,11 @@
 module Api
   class EventStreamsController < BaseController
+    def options
+      render_options(@req.collection, build_additional_fields)
+    end
+
+    def build_additional_fields
```
Comment (Fryguy) [inline app/controllers/api/event_streams_controller.rb:7]: This should be private.

File: spec/requests/event_streams_spec.rb [hunk 2]
```diff
@@ -162,4 +162,48 @@
...
+            },
+            "group_levels" => {
+              "critical" => "Critical",
+              "detail"   => "Detail",
+              "warning"  => "Warning"
+            }
+          },
+          "MiqEvent" => {
+            "description"  => "Policy Events",
+            "group_names"  => {},
```
Comment (Fryguy) [inline spec/requests/event_streams_spec.rb:201]: Is this right?

Comment (jrafanie) [reply spec/requests/event_streams_spec.rb:201]: Only if I'm not seeding, let me fix that.

File: spec/requests/event_streams_spec.rb [hunk 3]
```diff
@@ -162,4 +162,24 @@
...
+      MiqEventDefinitionSet.seed
+      options(api_event_streams_url)
+
+      expect_options_results(:event_streams)
+
+      body = response.parsed_body
+      expect(body["data"].keys).to eq(["timeline_events"])
+      expect(body["data"]["timeline_events"].keys.sort).to eq(%w[EmsEvent MiqEvent])
+      expect(body["data"]["timeline_events"]["EmsEvent"].keys.sort).to eq(%w[description group_levels group_names])
+      expect(body["data"]["timeline_events"]["EmsEvent"]["group_names"].keys.sort).to include("addition", "other")
```
Comment (jrafanie) [inline spec/requests/event_streams_spec.rb:177]: There's no need to test for all of them, just one from the event_handling and the default one "other"

File: spec/requests/event_streams_spec.rb [hunk 4]
```diff
@@ -162,4 +162,24 @@
...
+      expect(body["data"]["timeline_events"]["EmsEvent"]["group_levels"].keys.sort).to eq(%w[critical detail warning])
+
+      expect(body["data"]["timeline_events"]["MiqEvent"].keys.sort).to eq(%w[description group_levels group_names])
+      expect(body["data"]["timeline_events"]["MiqEvent"]["group_names"].keys.sort).to include("auth_validation", "other")
```
Comment (jrafanie) [inline spec/requests/event_streams_spec.rb:181]: same as above

File: spec/requests/querying_spec.rb [hunk 5]
```diff
@@ -966,7 +966,7 @@ def create_vms_by_name(names)
   describe 'OPTIONS /api/vms' do
     it 'returns the options information' do
       options(api_vms_url)
-      expect_options_results(:vms)
+      expect_options_results(:vms, {})
```
Comment (jrafanie) [inline spec/requests/querying_spec.rb:969]: Previously, this method defaulted to `{}` so we need to change the one caller that wasn't providing a hash and expected it to be defaulted.

**Review by Fryguy — APPROVED** [review]

**Pull request merged**